docs/proposals/vip-4.md

```diff
@@ -0,0 +1,217 @@
+
+## Scope
+
+This proposal is about full push jobs. Incremental pushes and nearline writes are out of scope. At the time of 
+submitting this proposal, it is undetermined whether this work will apply to stream reprocessing jobs. In terms of 
+priority, we care mostly about supporting full pushes from offline grids, and it may be fine to leave stream 
+reprocessing out of scope, although depending on the design details we choose, we may be able to support stream 
+reprocessing "for free" as well.
+
+The goal is for lifecycle hooks to achieve the following use cases:
+
+* Orchestrate how data is served in each region, including:
+  * Ensuring a minimum delay (e.g. 1 hour) between each region beginning to serve the new store-version.
+  * Delaying the swapping of a new store-version to be within some time of day (e.g. during "business hours").
+  * Performing custom health checks on the client applications to ensure that their key metrics are still healthy within
+    a region where a new store-version was swapped, before proceeding to more regions. Based on the outcome of this 
+    check:
+    * Having the ability to abort the swapping of a store-version to further regions.
+    * Having the ability to rollback to the previous store-version in regions that already swapped.
+* Trigger informational events in proprietary monitoring systems after important lifecycle milestones are completed.
+
+## Project Justification
+
+The cost of having global impact in the case of issues mentioned above is too high, and we would like to provide 
+first-class options to reduce the blast radius. Building this within Venice itself will make it easier to automate these
+methodologies, thus reducing toil for users.
+
+## Functional specification
+
+The proposed API for this functionality is described in code here:
+
+[//]: # (Got to remove the /venice prefix before pushing to main on linkedin/venice, otherwise links will be broken...)
+
+* [StoreLifecycleHooks](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleHooks.html), which is the main part of this 
+  proposal.
+* [StoreLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreLifecycleEventOutcome.html), which is the signal
+  returned by some hooks to indicate that a given step should proceed or abort.
+* [StoreVersionLifecycleEventOutcome](/venice/javadoc/com/linkedin/venice/hooks/StoreVersionLifecycleEventOutcome.html), which
+  is the signal returned by some other hooks which need more fine-grained control over the workflow. In addition to 
+  proceeding and aborting, this also provides the option to wait, which tells the hooks framework to try invoking the 
+  hook again later, and rollback, which tells the framework to rollback to the previous store-version in all regions.
+
+## Proposed Design
+
+The main design consideration is where to execute the hooks. At a high level, there are three options:
```

Comment:
Are we going to only choose one option? Just brainstorm here. Customers may choose to execute one individual hook per region for a life cycle event and after it run another hook that only executes once globally. Should this be supported? What's the max flexibility we want to provide?

Reply:
Yeah, good comments. A few thoughts on these points:

1. For the first iteration of this work, I think it would simplify things if we ran the hooks framework in only one component. My preference is to pick the parent controller, but I think VPJ might be fine too (different tradeoffs).
2. In the future, I think it is conceivable to expand the scope of hooks and that some of them could run inside other components, such as in the child controllers, or maybe even in the servers (I don't have a use case in mind for that, but theoretically, there could be...).
3. Regarding the flexibility of letting a hook implementer decide if a hook will run once per region or once globally, I think using either the parent controller or VPJ probably gets us closest to that... by letting one component handle all of the regions, then it may facilitate making certain types of decisions based on whether another regional hook already did something or not (although it involves maintaining state which could still be tricky, but not insurmountable...).

Reply:
There are different types of complexity if we choose different options.
The lifecycle management is stateful, which means if we choose to execute it in Controller, we may need to transfer the hook state during rebalance. Essentially, checkpointing and recovery will be required when changing Leader Controller for a given cluster, and if the hooks are executed in PushJob, I assume we don't need to handle such checkpointing/recovery logic since push job executor won't change during the lifetime of a certain run.

Reply:
Yeah that's right... state management is easier when using VPJ...

That being said, I am not sure that we have a use case that absolutely must have state.

Another aspect related to this is that if VPJ dies abruptly today (without being able to call its graceful kill/cancel function, which does happen sometimes) then as of today the push is capable of completing on its own, even without VPJ running (VPJ is only an observer). Whereas if we make VPJ the orchestrator, then VPJ dying can result in a hanging push (unless we adopt the VPJ heartbeat mechanism, which would be a good thing to have in and of itself, but causes scope creep in this project...).